### ISSUE TITLE
Specifying on_step=True in LightningModule.log only logs one value per epoch

### ISSUE THREAD

**Issue by rsokl** [PyTorchLightning/pytorch-lightning#4488]
Hello! I am hoping to understand the logging behavior that I am seeing.  

The docs for `LigthingModule.log` say:
- on_step – if True logs at this step
    - I think: "one value will be logged for each training step"
- on_epoch – if True logs epoch accumulated metrics
    - I think: "one accumulated value will be logged per epoch

Instead, both modes only log one value per epoch...

To be a bit more concrete, I expect for the following code to log one loss value for each training step

```python
def training_step(self, batch, batch_idx):
    # .... compute `loss`   
    self.log("loss", loss, on_step=True)
   return loss
```
Instead, I get one loss value logged per epoch.

## Diving a Little Deeper

I implemented the following simple logger:

```python
from pytorch_lightning.utilities import rank_zero_only
from pytorch_lightning.loggers import LightningLoggerBase
from collections import defaultdict

class ListLogger(LightningLoggerBase):
     logs: metric-name -> [(step0, v0), (step1, v1), ...]
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.data = defaultdict(list)
    
    @property
    def name(self): return "ListLogger"
    
    @property
    def experiment(self): return self.data

    @property
    def version(self): return "0.1"
    
    @rank_zero_only
    def log_hyperparams(self, params): pass
    
    @rank_zero_only
    def log_metrics(self, metrics, step):
        for k,v in metrics.items():
            self.data[k].append((step, v))
```

and modified my logging code to log loss both on-step and on-epoch

```python
def training_step(self, batch, batch_idx):
    # .... compute `loss`   
    self.log("on_step_train_loss", loss, on_step=True, on_epoch=False)
    self.log("on_epoch_train_loss", loss, on_step=False, on_epoch=True)
    return loss
```

The results are very confusing to me:

```python
>>> model.logger.data
defaultdict(list,
            {'on_step_train_loss': [(49, 1.4425818920135498),
              (99, 0.8221778869628906)],
             'epoch': [(49, 0), (61, 0), (99, 1), (123, 1)],
             'on_epoch_train_loss': [(61, 3.730947256088257),
              (123, 0.8831047415733337)]})
```

Can someone please explain what `on_step=True` means, if it doesn't mean "log one value per step"? Is there some accumulation going on here? What is the significance of steps `49` and `99` - the steps where `on_step=True` seems to actually trigger logging?

**Comment by s-rog** [comment]
https://pytorch-lightning.readthedocs.io/en/stable/trainer.html#log-every-n-steps

You didn't post your trainer code but this might be the source of your difficulties?

**Comment by rsokl** [comment]
Ah yep! I can't believe I missed this: https://pytorch-lightning.readthedocs.io/en/stable/logging.html#control-logging-frequency

thank you!

**Comment by asude-sony** [comment]
None of these links work anymore
